peer/connector.go

```diff
@@ -80,48 +83,91 @@ func (conn *connector) initialise(privateKey []byte, publicKey []byte, connect [
-			errX = err
-			goto fail
-		}
+		wg.Add(1)
 
-		// prevent connection to self
-		if bytes.Equal(publicKey, serverPublicKey) {
-			errX = fault.ErrConnectingToSelfForbidden
-			log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, errX)
-			goto fail
-		}
+		// start new goroutine for each connection
+		go func(conn *connector, c Connection, i int, wg *sync.WaitGroup, ch chan error) {
 
-		client, err := upstream.New(privateKey, publicKey, connectorTimeout)
-		if nil != err {
-			log.Errorf("client[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
+			// error function call
+			errF := func(wg *sync.WaitGroup, ch chan error, e error) {
+				ch <- e
+				wg.Done()
+			}
 
-		conn.staticClients[i] = client
-		globalData.connectorClients = append(globalData.connectorClients, client)
+			// for canonicaling the error
+			canonicalErrF := func(c Connection, e error) error {
+				return fault.GenericError(fmt.Sprintf("client: %q error: %s", c.Address, e))
+			}
 
-		err = client.Connect(address, serverPublicKey)
-		if nil != err {
-			log.Errorf("connect[%d]=%q  error: %s", i, address, err)
-			errX = err
-			goto fail
-		}
-		log.Infof("public key: %x  at: %q", serverPublicKey, c.Address)
+			address, err := util.NewConnection(c.Address)
+			if nil != err {
+				log.Errorf("client[%d]=address: %q  error: %s", i, c.Address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+			serverPublicKey, err := hex.DecodeString(c.PublicKey)
+			if nil != err {
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			// prevent connection to self
+			if bytes.Equal(publicKey, serverPublicKey) {
+				err := fault.ErrConnectingToSelfForbidden
+				log.Errorf("client[%d]=public: %q  error: %s", i, c.PublicKey, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			client, err := upstream.New(privateKey, publicKey, connectorTimeout)
+			if nil != err {
+				log.Errorf("client[%d]=%q  error: %s", i, address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+
+			conn.Lock()
+			conn.staticClients[i] = client
+			conn.Unlock()
+
+			globalData.connectorClients = append(globalData.connectorClients, client)
+
+			err = client.Connect(address, serverPublicKey)
+			if nil != err {
+				log.Errorf("connect[%d]=%q  error: %s", i, address, err)
+				errF(wg, ch, canonicalErrF(c, err))
+				return
+			}
+			log.Infof("public key: %x  at: %q", serverPublicKey, c.Address)
+			wg.Done()
+
+		}(conn, c, i, &wg, errCh)
+	}
+
+	conn.log.Debug("waiting for all static connections...")
+	wg.Wait()
+
+	// drop error channel for getting all errors
+	errXs := make([]error, 0)
+	for len(errCh) > 0 {
+		errXs = append(errXs, <-errCh)
+	}
+
+	// error code for goto fail
+	errX := error(nil)
+
+	if len(errXs) > 0 {
+		errX = compositeError(errXs)
+		conn.log.Debugf("error: %s", errX)
+		goto fail
```

Comment:
These code block is quite big, any idea to have tests?

Reply:
Because this function is private, so i think if we want to test it, we need to test the caller. In this case we need to test the `peer/setup.go - Initialise()`. Currently i don't know how to test it, just run the bitmarkd and verify by the log.